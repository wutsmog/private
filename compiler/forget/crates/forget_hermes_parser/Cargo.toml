[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }