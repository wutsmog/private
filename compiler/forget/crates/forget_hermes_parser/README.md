# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.