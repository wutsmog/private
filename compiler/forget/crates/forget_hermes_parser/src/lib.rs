mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
