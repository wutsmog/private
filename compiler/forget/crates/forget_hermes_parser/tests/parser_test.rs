use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
