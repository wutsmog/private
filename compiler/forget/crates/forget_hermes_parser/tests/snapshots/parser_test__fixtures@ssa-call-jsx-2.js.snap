---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/ssa-call-jsx-2.js
---
Input:
// @Pass runMutableRangeAnalysis
function foo() {}

function Component(props) {
  const a = [];
  const b = {};
  foo(a, b);
  if (foo()) {
    let _ = <div a={a} />;
  }
  foo(a, b);
  return <div a={a} b={b} />;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "foo",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 42,
          "end": 45
        }
      },
      "params": [],
      "body": {
        "type": "BlockStatement",
        "body": [],
        "loc": null,
        "range": {
          "start": 48,
          "end": 50
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 33,
        "end": 50
      },
      "loc": null,
      "range": {
        "start": 33,
        "end": 50
      }
    },
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 61,
          "end": 70
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 71,
            "end": 76
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "a",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 88,
                    "end": 89
                  }
                },
                "init": {
                  "type": "ArrayExpression",
                  "elements": [],
                  "loc": null,
                  "range": {
                    "start": 92,
                    "end": 94
                  }
                },
                "loc": null,
                "range": {
                  "start": 88,
                  "end": 94
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 82,
              "end": 95
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "b",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 104,
                    "end": 105
                  }
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [],
                  "loc": null,
                  "range": {
                    "start": 108,
                    "end": 110
                  }
                },
                "loc": null,
                "range": {
                  "start": 104,
                  "end": 110
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 98,
              "end": 111
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "Identifier",
                "name": "foo",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 114,
                  "end": 117
                }
              },
              "arguments": [
                {
                  "type": "Identifier",
                  "name": "a",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 118,
                    "end": 119
                  }
                },
                {
                  "type": "Identifier",
                  "name": "b",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 121,
                    "end": 122
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 114,
                "end": 123
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 114,
              "end": 124
            }
          },
          {
            "type": "IfStatement",
            "test": {
              "type": "CallExpression",
              "callee": {
                "type": "Identifier",
                "name": "foo",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 131,
                  "end": 134
                }
              },
              "arguments": [],
              "loc": null,
              "range": {
                "start": 131,
                "end": 136
              }
            },
            "consequent": {
              "type": "BlockStatement",
              "body": [
                {
                  "type": "VariableDeclaration",
                  "kind": "let",
                  "declarations": [
                    {
                      "type": "VariableDeclarator",
                      "id": {
                        "type": "Identifier",
                        "name": "_",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 148,
                          "end": 149
                        }
                      },
                      "init": {
                        "type": "JSXElement",
                        "openingElement": {
                          "type": "JSXOpeningElement",
                          "name": {
                            "type": "JSXIdentifier",
                            "name": "div",
                            "loc": null,
                            "range": {
                              "start": 153,
                              "end": 156
                            }
                          },
                          "attributes": [
                            {
                              "type": "JSXAttribute",
                              "name": {
                                "type": "JSXIdentifier",
                                "name": "a",
                                "loc": null,
                                "range": {
                                  "start": 157,
                                  "end": 158
                                }
                              },
                              "value": {
                                "type": "JSXExpressionContainer",
                                "expression": {
                                  "type": "Identifier",
                                  "name": "a",
                                  "typeAnnotation": null,
                                  "loc": null,
                                  "range": {
                                    "start": 160,
                                    "end": 161
                                  }
                                },
                                "loc": null,
                                "range": {
                                  "start": 159,
                                  "end": 162
                                }
                              },
                              "loc": null,
                              "range": {
                                "start": 157,
                                "end": 162
                              }
                            }
                          ],
                          "selfClosing": true,
                          "loc": null,
                          "range": {
                            "start": 152,
                            "end": 165
                          }
                        },
                        "children": [],
                        "closingElement": null,
                        "loc": null,
                        "range": {
                          "start": 152,
                          "end": 165
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 148,
                        "end": 165
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 144,
                    "end": 166
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 138,
                "end": 170
              }
            },
            "alternate": null,
            "loc": null,
            "range": {
              "start": 127,
              "end": 170
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "Identifier",
                "name": "foo",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 173,
                  "end": 176
                }
              },
              "arguments": [
                {
                  "type": "Identifier",
                  "name": "a",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 177,
                    "end": 178
                  }
                },
                {
                  "type": "Identifier",
                  "name": "b",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 180,
                    "end": 181
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 173,
                "end": 182
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 173,
              "end": 183
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "JSXElement",
              "openingElement": {
                "type": "JSXOpeningElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": {
                    "start": 194,
                    "end": 197
                  }
                },
                "attributes": [
                  {
                    "type": "JSXAttribute",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "a",
                      "loc": null,
                      "range": {
                        "start": 198,
                        "end": 199
                      }
                    },
                    "value": {
                      "type": "JSXExpressionContainer",
                      "expression": {
                        "type": "Identifier",
                        "name": "a",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 201,
                          "end": 202
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 200,
                        "end": 203
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 198,
                      "end": 203
                    }
                  },
                  {
                    "type": "JSXAttribute",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "b",
                      "loc": null,
                      "range": {
                        "start": 204,
                        "end": 205
                      }
                    },
                    "value": {
                      "type": "JSXExpressionContainer",
                      "expression": {
                        "type": "Identifier",
                        "name": "b",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 207,
                          "end": 208
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 206,
                        "end": 209
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 204,
                      "end": 209
                    }
                  }
                ],
                "selfClosing": true,
                "loc": null,
                "range": {
                  "start": 193,
                  "end": 212
                }
              },
              "children": [],
              "closingElement": null,
              "loc": null,
              "range": {
                "start": 193,
                "end": 212
              }
            },
            "loc": null,
            "range": {
              "start": 186,
              "end": 213
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 78,
          "end": 215
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 52,
        "end": 215
      },
      "loc": null,
      "range": {
        "start": 52,
        "end": 215
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 33,
    "end": 215
  }
}
